feat!: tag categories

CyanVoxel opened the pull request:
### Replaces "Tag Fields" with Tag Categories.

Instead of tags needing to be added to a Tag field type such as "Meta Tags", "Content Tags", or simply the "Tags" field, tags are now added directly to file entries with no intermediary step. While tag field types offered a method to further organize tags, it was cumbersome, inflexible, and simply not fully fleshed out. Tag Categories offer all of the previous (intentional) functionality while greatly increasing the ease of use and customization.

### How it works
Tags now have an additional boolean property that states whether or not a tag is marked as a "category" or not. Tags marked as categories have no functional difference in the backend but are treated specially by frontend views. Category tags take on the appearance similar to the previous tag fields while automatically showing all tags inheriting from that tag underneath its own group. This means that duplicates of tags can appear on entries if they inherit from multiple categories, however this is by design. Any tags not inheriting from a category tag will simply show under a default "Tag" field-like section.

The previous built-in tags, "Favorite" and "Archived", now inherit from a new built-in "Meta Tags" tag that's marked as a category by default. This means that the appearance of these tags is effectively unchanged from how they're added via the badge buttons - however the organization is now fully automatic, universal across all tag adding sources, and fully customizable by the user.

This *does* mean that existing user tag field configurations will be changed. Instead of tags being organized into field types on an per-entry basis, tags themselves determine their organizational layout via the new category flag. Any tags (not currently inheriting from either the "Favorite" or "Archived" tags) will be shown under the default "Tags" header upon migrating from v9.4 libraries. If the user wishes to permanently mark some of their own tags as "Meta Tags" they can easily do that by having these tags inherit from the new built-in "Meta Tags" tag. And if they'd rather set some of their own existing tags as categories (i.e. "People", "Places", "Characters") then they are free to do so as well!

<img width="279" alt="Screenshot 2025-01-04 at 04 23 43" src="https://github.com/user-attachments/assets/0b92eca5-db8f-4e3e-954b-1b4f3795f073" />

### Related Changes

A major component of this PR has been a comprehensive refactor of the preview panel. The panel has been broken up in the following ways:
- `PreviewPanel` has been made a lightweight container widget
- Image/Media Previews have been broken off into `PreviewThumb`
- File Attributes have been broken off into `FileAttributes`
- Field Containers have been broken off into `FieldContainers`
- The "Recent Libraries" panel has been converted into a submenu under "File -> Open Recent"

The following features have not been re-implemented for the following reasons:
- The dedicated "Recent Libraries" panel: I have opted to move this to the File menu to free up valuable screen real estate used by tags and fields in the preview panel, and to convert this feature into a more mainstream implementation. I wouldn't mind if someone re-implemented the panel widget (with the option to disable) if they wished, however I have no interest to refactor this myself. 
- Mixed Selection updating and deleting: While adding tags and fields to multiple entries at once is still possible, editing and updating is currently not in this PR. Since these features already have numerous issues with them I've opted to disable them here while saving a more comprehensive rework of these (including addressing #337) for a future PR. 

***
As a bonus with this approach, this PR fixes numerous issues present with field editing and rendering, both documented and undocumented since the SQL switch. Currently this PR is set to close #92, close #249, close #18, close #321, and close #416.

Computerdores commented:
While you are already chaning the DB, you might want to also fix the `tag_subtags` table.
(I would rename it to something like `tag_inheritance` so that `subtag` isn't in the name anymore and most importantly swap the column names so they are actually correct)

CyanVoxel commented:
> While you are already changing the DB, you might want to also fix the `tag_subtags` table. (I would rename it to something like `tag_inheritance` so that `subtag` isn't in the name anymore and most importantly swap the column names so they are actually correct)

I would **love** to do this 👍

Computerdores commented on tagstudio/src/qt/widgets/migration_modal.py:
```diff
@@ -553,70 +553,68 @@ def sanitize_json_field(value):
-                            sql_entry.id,
-                            sf.type.key,
-                            sanitize_field(session, sql_entry, sf.value, sf.type.type, sf.type_key),
+                    if sf.type.type.value not in {6, 7, 8}:
```
same thing about the magic numbers here

Computerdores commented on tagstudio/src/qt/widgets/migration_modal.py:
```diff
@@ -553,70 +553,68 @@ def sanitize_json_field(value):
-                        content_tags_count += 1
-                        merged_content_tags = merged_content_tags.union(value or [])
-                    elif key == _FieldID.TAGS.name:
+                    if int_key in {6, 7, 8}:
```
magic numbers - see above

Computerdores commented on tagstudio/src/qt/widgets/migration_modal.py:
```diff
@@ -655,14 +653,17 @@ def check_subtag_parity(self) -> bool:
         with Session(self.sql_lib.engine) as session:
             for tag in self.sql_lib.tags:
+                if tag.id in range(0, 1000):
+                    break
```
why `break` here? shouldn't it `continue`?
(same for the other loops in `check_alias_parity`, `check_shorthand_parity`, and `check_color_parity`)

CyanVoxel replied:
This is temporary code that's skipping the JSON builtin tags as the migration system hasn't been fixed to work with those yet. This code is purposely not migrating those IDs temporarily while a the default tag migration system is rebuilt for the changes here.

CyanVoxel replied:
Shoot I think you're right, *now* these should be `continue`s...
I mentioned this in the discord too, but yeah one thing that slipped by me was carrying over preexisting modified built-in tags and combining those with the new default data (in this case I think it's just parenting to meta/2). Actually when I word it like that, I could just get rid of these checks and just have a little hardcoded addition of "2" (rather, the named constant) to the json code before migrating it.

Computerdores replied:
Yeah given the size of this PR I was already suspecting you might just have missed that

Computerdores reviewed (commented):
Had some free time and wanted to get ahead of the curve a little bit as this PR seems to be shaping up to be a big one.
Feel free to ignore for now, the PR is still a draft after all.

CyanVoxel commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -193,11 +207,15 @@ def migrate_json_to_sqlite(self, json_lib: JsonLibrary):
-                        value=v,
-                    )
+                    # Old tag fields get added as tags
+                    if k in {6, 7, 8}:
```
I'll definitely store them in a more human readable way, this was just to make sure the concept worked

CyanVoxel commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -256,14 +272,13 @@ def open_sqlite_library(
-                except IntegrityError:
-                    # default tags may exist already
-                    session.rollback()
+            tags = get_default_tags()
+            try:
+                session.add_all(tags)
+                session.commit()
+            except IntegrityError:
+                # default tags may exist already
+                session.rollback()
```
Still a work in progress, I haven't worked out a way to add the default tags while not stepping on existing libraries while also allowing JSON libraries with modified defaults to be migrated. This code is from main just moved around while figuring this out

python357-1 replied:
Can we not rely on the ids being the same every time?

CyanVoxel replied:
Maybe I was referring to a problem that this doesn't effect yet, since yeah I don't think this overwrites existing tags. The `add_default_data` flag was only added in the migration PR because the previous behavior screwed with getting a blank library to import the JSON tags into. Otherwise this is the same behavior as on main for any other library

CyanVoxel commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -157,14 +167,18 @@ def migrate_json_to_sqlite(self, json_lib: JsonLibrary):
-                    shorthand=tag.shorthand,
-                    color=TagColor.get_color_from_str(tag.color),
+            if tag.id == TAG_ARCHIVED or tag.id == TAG_FAVORITE:
+                # Update built-in
```
Could be, but I'm aware that the migration code for default tags is still incomplete in this draft

CyanVoxel replied:
Just realized I need to scroll up in order to access the reply box for this, but I've got an update on this situation in my comment here: https://github.com/TagStudioDev/TagStudio/pull/655#issuecomment-2585261121

CyanVoxel commented on tagstudio/src/qt/ts_qt.py:
```diff
@@ -135,8 +135,9 @@ def __init__(self, backend, args):
-        self.frame_content = []
         self.filter = FilterState.show_all()
+        self.frame_content: list[Entry] = []
+        self.filter = FilterState().show_all()
```
This seems like an older line that was accidentally left in, seeing as 138 contains the suggested change

CyanVoxel commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -71,24 +70,35 @@ def slugify(input_string: str) -> str:
 
 
 def get_default_tags() -> tuple[Tag, ...]:
+    meta_tag = Tag(
+        id=TAG_META,
+        name="Meta Tags",
+        aliases={TagAlias(name="Meta"), TagAlias(name="Meta Tag")},
+        is_category=True,
+    )
```
I want the built-in category to still be called "Meta Tags" since that will appear as a section header like fields do and not as standalone tags, and just having a section called "Meta" where tags appear with no indication that they're tags would come off as strange to me

Tishj commented:
Couldn't resist so I tested it a little, when I have a tag `A` that is a category, and I have a tag `B` that has `A` as parent, it will show up under the `A` category

Then when I create a tag `C` that has `B` as parent, `C` will not show up under the `A` category, is that expected behavior?

```
A
└── B      // shows up under A
    └── C  // does not show up under A
```

Computerdores commented:
> Couldn't resist so I tested it a little, when I have a tag `A` that is a category, and I have a tag `B` that has `A` as parent, it will show up under the `A` category
> 
> Then when I create a tag `C` that has `B` as parent, `C` will not show up under the `A` category, is that expected behavior?
> 
> ```
> A
> └── B      // shows up under A
>     └── C  // does not show up under A
> ```

I believe Cyan is currently working on fixing that

Tishj commented:
Awesome! 👍

CyanVoxel commented:
> Couldn't resist so I tested it a little, when I have a tag `A` that is a category, and I have a tag `B` that has `A` as parent, it will show up under the `A` category
> 
> Then when I create a tag `C` that has `B` as parent, `C` will not show up under the `A` category, is that expected behavior?
> 
> ```
> A
> └── B      // shows up under A
>     └── C  // does not show up under A
> ```

You just happened to catch me as I (hopefully) finally just got this working in the latest commit! 😁

Tishj commented:
I checked out the last commit and it works as expected, but now I'm finding situations where I wouldn't really want this to happen :/ So it's slightly more nuanced than this, let me try to give an example:

Lets say I have a library of books, and these books have characteristics, the themes are a category, the setting is a category, the time period as well
And then some authors are tags, because some authors are known for having the same recurring theme or setting in their books

As a result, the `Theme`, `Setting` and `Time Period` categories now show `John Doe` because John Doe is known for those tags, so they are bundled under John Doe

Might be a bad example, but it's related to bundling of tags under one tag, and the tags that are bundled are part of several categories

CyanVoxel commented:
> I checked out the last commit and it works as expected! But now I'm finding situations where I wouldn't really want this to happen :/ So it's slightly more nuanced than this.
> ### Problematic Case
> 
> Lets say I have a library of books, and these books have characteristics, the themes are a category, the setting is a category, the time period as well And then some authors are tags, because some authors are known for having the same recurring theme or setting in their books
> 
> As a result, the `Theme`, `Setting` and `Time Period` categories now show `John Doe` because John Doe is known for featuring specific themes and all their books feature a specific time period, so they are bundled under John Doe
> 
> Might be a bad example, but it's related to bundling of tags under one tag, and the tags that are bundled are part of several categories

I think this is more so an issue that arises with that approach to tagging rather than the category system. Parent Tags are intended to be used in an inheritance type of relationship with other tags. This works well for books that *are* certain genres, but not for describing someone like "John Doe" as "Science Fiction" because he writes in that genre often. "John Doe" may have some association with the genre, but he *isn't* that genre in the same way that a book may be a genre, or how another genre such as "Hard Sci-Fi" may also *be* and inherit from "Science Fiction". Having genre parent tags on authors themselves would also break as soon as you came across a book by them that no longer fell under that genre - there would no longer be a way to get rid of that genre from the file entry.

***Note:** For 9.6 I'm planning to introduce a new type of subtag in addition to Parent Tags called "Component Tags" which will use a composition/HAS relationship instead of the inheritance/IS relationship that Parent Tags use. I'm not sure how I want this to work with the tag category system yet, so I'm keeping my eyes out for situations that may help inform this behavior.*

> One easy-ish(?) solution would be to not display tags under a category if the tag itself is a category Then the author tags could be marked as categories to avoid showing them under the categories of their parent tags

In this case wouldn't you be marking "John Doe" and every other author as categories? That wouldn't be great from a visual perspective with each author displaying under their own self-titled category, and also wouldn't work from an organizational perspective as those authors are not categories themselves but rather entities who fall under one or more categories such as, well, *authors*

> Another idea would be to add another checkbox that disables categorization for a tag, so they only show under the `Tags` section.

This would work, however I (perhaps stubbornly) believe that if a tag inherits from a tag marked as a category then it *should* belong under that category. If a case such as the one you described with the genres and authors arise, I would consider that an issue with the tagging approach rather than the category system itself. Same if there was an instance where a separate category didn't make sense for a group of tags - in that case, just don't display that tag as a category and instead rethink the nature of how you would like certain tags to be organized.  I'm open to hearing more cases against this if they arise, though. Life is full of more edge cases than I can ever encounter just by my own experiences, after all.

> A third idea could be to disassemble the tag, the category would only contain the direct child, letting the further descendants be displayed under `Tags` Then the further descendant can explicitly inherit from the grandparent (category) tag to be included under that category (so `C `from the earlier example would need to add `A` as parent)

This would also unfortunately undermine much of the automatic inheritance provided by the tag category system. For example I have several tags such as "Media Type" for photos vs illustrations which I keep as a child of "Meta Tags". This allows me to have `C` tags like "Photo", "Video", "Drawing", "3D Render", etc be organized under as many `B`-like tags as I want or need, while still appearing under the category for `A` (Meta Tags) without me needing to manually mark each of these `C` tags.

Tishj commented:
Thanks for the detailed reply!

I understand that these issues arise from the way I'm treating these tag hierarchies, I think the described component tags would solve my problem, because essentially I want to bundle an assortment of tags under a new name, so I don't have to assign all of the bundled tags separately and manually to an entry.

Perhaps a more clear example would be something like a `Generation` category, where the author (John Doe) gets assigned a tag like `Millenial`/`Boomer` that inherits from `Generation`.
As a result `John Doe` would show up under the `Generation` category, which is weird

CyanVoxel commented:
> > In this case wouldn't you be marking "John Doe" and every other author as categories? That wouldn't be great from a visual perspective with each author displaying under their own self-titled category
> 
> At least in the current state of the branch, I have found that the tag that is marked as the category does not show up under its own category when assigned to an entry, they display under `Tags`, only the tags that inherit from them are shown under the category. Which is why I proposed that idea, since there isn't a tag that inherits from John Doe, the John Doe category would never appear

I see, I just happened to see and fix that bug (88b886c1dfa96919db2446c07e32811c0eb6f871) after your message haha. It slipped past me while I was working on this latest approach to recursive categories, but it should be fixed now where category tags themselves show under their own categories.

CyanVoxel commented:
> Thanks for the detailed reply!
> 
> I understand that these issues arise from the way I'm treating these tag hierarchies, I think the described component tags would solve my problem, because essentially I want to bundle an assortment of tags under a new name, so I don't have to assign all of the bundled tags separately and manually to an entry.
> 
> Perhaps a more clear example would be something like a `Generation` category, where the author (John Doe) gets assigned a tag like `Millenial`/`Boomer` that inherits from `Generation`. As a result `John Doe` would show up under the `Generation` category, which is weird

This... might a good example of an issue with the category system. Another one I realized could be having a "Movie" tag and a "Character" tag, with a tag "Shrek" (Character) inheriting from "Shrek" (Movie) and "Character". If "Character" is marked as a category, all is fine. But if you *also* wanted "Movie" as a category, then "Shrek" (Character) would also show up as a "Movie" because it *is* a Shrek Movie franchise entity. Not ideal. But that also comes with a couple quirks of the tagging method:
1. In that library, we're using the Character tags such as Shrek to contain movie information, *not* movie tags on entries themselves. We don't need to add tags for the Shrek movie since Shrek the character tag holds that inheritance relationship.
*Although*, this falls apart if you have other types of entries in your library that don't use characters and need to inherit straight from "Movie", while making that a category - which I think is a valid use case that breaks the tag category system as it stands... The same would be true for your "Generation" tag, if you wanted to use that for other non-author types of file entries. 

2. While having Shrek (Movie) be a component tag on Shrek (Character) *would* properly prevent the character from showing up under the "Movie" category, it would also break what is intended to be an inheritance relationship with the Movie. If a tag such as, oh I really don't know, "Shrek Backpack" included the Shrek (Character) tag, it would *not* show up in searches for Shrek (Movie) due to component tags not being planned to be passed down like inherited tags (i.e. Shrek (Character) can not substitute for its component tag, Shrek (Movie); Only for its parent tags like "Character").
A user-side fix to this would be to add the Shrek (Movie) tag to every entity that's included in that movie franchise, but I feel that's going against what makes TagStudio easy to use with automatic tag inheritance.
Another "fix" would be to have a Shrek (Franchise) tag from which both Shrek (Movie) and Shrek (Character) inherit from - which would allow both Movie and Character to properly show their own children. Actually I feel this is the proper way to describe this relationship regardless of tag categories. Buuuuut, the same issue arises as soon as you want "Franchise" to be its own category, too...

I'll continue to give this some thought in the hopes that there's a simple and general solution here either in tagging methods themselves, with 9.6 component tags, or simply with a tweak to how tag categories operate. My worry is that there won't be a way to fix this without doing something much more tailored and involved to category tags, such as having an extra subtag box where users can list which potential child tags this category should not show. That may function but I would consider that a very messy fix that would also have to extend as an option across all types of tags. Of course in the meantime, I'm very open to other proposals or scenarios that may help shed some more light on this.

Computerdores commented on tagstudio/src/qt/widgets/preview/field_containers.py:
```diff
@@ -0,0 +1,520 @@
+                if cluster_map.get(p_id) is None:
+                    cluster_map[p_id] = set()
+                # If the p_tag has p_tags of its own, recursively link those to the original Tag.
+                if tag.id not in cluster_map[p_id]:
+                    cluster_map[p_id].add(tag.id)
```
there is a typo here: `tag.id` should be `tag_id` (or at least I can't find where `tag` was defined in this method)
```suggestion
                if tag_id not in cluster_map[p_id]:
                    cluster_map[p_id].add(tag_id)
```

Computerdores commented on tagstudio/src/qt/widgets/preview/field_containers.py:
```diff
@@ -0,0 +1,520 @@
+        logger.info("[FieldContainers] Cluster Map", cluster_map=cluster_map)
+
+        # Initialize all categories from parents.
+        tags_ = {self.lib.get_tag(x) for x in exhausted}
```
`exhausted` just ends up being `cluster_map.keys()` plus the ids from `tags` (which shouldn't be added anyways since that might lead to empty categories being shown, right?) - so replacing refernces to `exhausted` with `cluster_map.keys()` should be fine and simplify the code

CyanVoxel replied:
Replacing `exhausted` references with `cluster_map.keys()` (or specifically, replacing `tags_ = {self.lib.get_tag(x) for x in exhausted}` with `tags_ = {self.lib.get_tag(x) for x in cluster_map}` has the side effect of not showing a category under its own category in the event it's added to an entry - but only if alone.

Using `tags_ = {self.lib.get_tag(x) for x in exhausted}`:
![image](https://github.com/user-attachments/assets/ee531347-1ccc-42f7-93ae-73d8daf82072)

Using `tags_ = {self.lib.get_tag(x) for x in cluster_map}`
![image](https://github.com/user-attachments/assets/2eae886c-74c1-485c-9a83-5e5cd3e0d260)
![image](https://github.com/user-attachments/assets/9ed73a49-e4fa-4f78-bd5a-87af80a17085)
This change also causes the `test_custom_tag_category` test to fail for seemingly the same reason.

Computerdores replied:
nvm then

Computerdores replied:
True I didn't think of that - however should a category even be shown under itself? Because Color isn't a Color, Character isn't a Character, etc.

CyanVoxel replied:
Would there be alternatives other than showing the tag as a blank category or showing the category under the general "Tags" section? It may be a bit quirky, but I feel this approach makes the most sense out of the alternatives - and in *most* cases you aren't going to want a category tag on an entry by itself, you'd instead add a tag that inherits from that category instead. Personally I prefer the category tags showing under their own categories as it makes it easier to spot and replace with more appropriate ones.

Computerdores replied:
> Personally I prefer the category tags showing under their own categories as it makes it easier to spot and replace with more appropriate ones.

Yeah honestly that is fair enough, I can live with both anyways - do as you like

CyanVoxel replied:
I'd also be open to making the behavior configurable down the line, if people wish. There's definitely some edge cases to contemplate about this system, notably the whole "shrek backpack" scenario...

Computerdores commented on tagstudio/src/qt/ts_qt.py:
```diff
@@ -972,44 +974,61 @@ def _init_thumb_grid(self):
         sa.setWidget(self.flow_container)
 
-    def select_item(self, grid_index: int, append: bool, bridge: bool):
+    def select_item(self, item_id: int, append: bool, bridge: bool):
```
I kind of dislike the name of this method since calling it sometimes selects and sometimes deselects an item based on whether it is already selected, but the name would suggest that it is always selected

Tishj replied:
Based on that description, perhaps `toggle_item_select` ?

Computerdores reviewed (commented):
Notes:
- Given that the new ui files each have hundreds of lines of ui code that isn't directly related to the core functionality I have tried to skim those parts and focus only on the more acutely relevant parts. This PR is already big enough without that after all.
- completely commenting out some of the tests is a bit iffy, but I think it'll be fine

CyanVoxel commented on tagstudio/src/core/library/alchemy/visitors.py:
```diff
@@ -51,12 +56,18 @@ def visit_and_list(self, node: ANDList) -> ColumnExpressionArgument:
                 match term.type:
                     case ConstraintType.TagID:
-                        tag_ids.append(int(term.value))
+                        try:
+                            tag_ids.append(int(term.value))
+                        except ValueError:
+                            logger.error(
+                                "[SQLBoolExpressionBuilder] Could not cast value to an int Tag ID",
+                                value=term.value,
+                            )
```
I believe this happens when you have a search that starts with `tag_id: <id>` and then is followed by a space and a string, for example `tag_id: 2000 favorite`. Probably something to catch on the parser side but it came up when working on this so I plugged it here

Computerdores replied:
I would like to fix that at some point, but that is definitely out of scope for this PR

CyanVoxel commented on tagstudio/src/core/palette.py:
```diff
@@ -25,7 +25,8 @@ class UiColor(IntEnum):
     RED = 3
     GREEN = 4
-    PURPLE = 5
+    BLUE = 5
+    PURPLE = 6
```
These values for the UI are only ever read from at runtime and aren't stored anywhere on disk, so there shouldn't be any technical side effects. I don't know why they're enums now, but I did this to better organize the list order by color while they're still here. Imo I would check for a "purple" color in-between red and blue

CyanVoxel commented on tagstudio/src/qt/ts_qt.py:
```diff
@@ -972,44 +974,61 @@ def _init_thumb_grid(self):
-                self.selected.append(grid_index)
-                self.item_thumbs[grid_index].thumb_button.set_selected(True)
+            if item_id not in self.selected:
+                self.selected.append(item_id)
+                for it in self.item_thumbs:
+                    if it.item_id == item_id:
+                        it.thumb_button.set_selected(True)
             else:
-                self.selected.remove(grid_index)
-                self.item_thumbs[grid_index].thumb_button.set_selected(False)
+                self.selected.remove(item_id)
+                for it in self.item_thumbs:
+                    if it.item_id == item_id:
+                        it.thumb_button.set_selected(False)
```
Maybe it's a holdover from my thinking in previous TS versions, but if there's ever a case where there's the same entry or item shown on screen then this would update it everywhere it's shown - for example maybe some specific grouping views. Maybe it's over-engineering on my part though.

CyanVoxel replied:
Actually I may have been thinking of the general selection case further down, not the append case.
But now that the grid indices of item_thumbs aren't being stored anymore this is the *one* area where it's negatively affected compared to the improvements everywhere else. *Personally* I don't notice a performance hit, but if it becomes an issue then we can do something like add an ID to index map and reference from that.

CyanVoxel commented on tagstudio/src/qt/ts_qt.py:
```diff
@@ -1106,18 +1125,24 @@ def update_thumbs(self):
-        for entry, item_thumb in zip_longest(self.frame_content, self.item_thumbs):
-            if not entry:
+        logger.info("[QtDriver] Loading Entries...")
+        entries: list[Entry] = list(self.lib.get_entries_full(self.frame_content))
```
You're right, I don't think it needs to be done - there just isn't a method to select a list of specific entries without joins at the moment. I tried combining `get_entries_full()`'s ability to pass ids with `get_entries()`'s ability to specify joins but couldn't get something working. Really I think that `get_entry()`, `get_entry_full()`, `get_entries()`, and `get_entries_full()` could all probably be merged into a single method - I just haven't been successful in my attempts so far...

CyanVoxel replied:
Added a TODO comment about this in 5e1e4c57a73dfaef7f6ae9ff233376b12a775704

CyanVoxel commented on tagstudio/src/qt/widgets/migration_modal.py:
```diff
@@ -498,136 +501,79 @@ def color_value_conditional(self, old_value: int | str, new_value: int | str) ->
-                return set(tags) if tags else None
-            else:
-                return value if value else None
+        def sanitize_field(entry: Entry, value, type, type_key):
+            return value if value else None
```
Turns out this is still needed in order to compare against `None` cases from the JSON side. While that side could be modified to return `''` strings instead of `None`, I feel `None` in this case helps reduce ambiguity both in the code and in discrepancy logs.

Here's the discrepancies returned by removing this method on the built-in migration test library:
```
[Field Comparison]:
OLD (JSON):[(1, 'ARTIST', None), (1, 'AUTHOR', None), (1, 'COMMENTS', None), (1, 'COMPOSER', None), (1, 'DESCRIPTION', None), (1, 'GUEST_ARTIST', None), (1, 'NOTES', None), (1, 'PUBLISHER', None), (1, 'TITLE', None), (1, 'URL', None)]
set()
NEW  (SQL):[(1, 'ARTIST', ''), (1, 'AUTHOR', ''), (1, 'COMMENTS', ''), (1, 'COMPOSER', ''), (1, 'DESCRIPTION', ''), (1, 'GUEST_ARTIST', ''), (1, 'NOTES', ''), (1, 'PUBLISHER', ''), (1, 'TITLE', ''), (1, 'URL', '')]
set()
```

Tishj commented:
I didn't want to report it before because the PR was a draft, but with a build from this PR I can't open files from tagstudio in a library migrated from 9.4 (migration also done with this build)

I can click the option, but it doesn't do anything
It's like it can't find the path or something? and just silently fails

CyanVoxel commented:
> I didn't want to report it before because the PR was a draft, but with a build from this PR I can't open video files from tagstudio in a library migrated from 9.4 (migration also done with this build)
> 
> I can click the option, but it doesn't do anything It's like it can't find the path or something? and just silently fails

This happens if you open a preexisting migrated library that was created with an experimental version from the `main` branch. Main is currently unstable as noted by the banner at the top of the README, CONTRIBUTING.md, as well as the "EXPERIMENTAL" warning in the program's title bar. Libraries migrated or created with these experimental versions before the DB schema is finalized are not officially supported - only versions across released builds are supported while we transition to SQL. It would be too much work otherwise to build complete migration tools and checks for each database change that comes in while we try fixing issues with the SQL schema. Since this PR drastically changes elements of the schema (and is continuing to tweak things during review) you'll likely have to re-migrate your JSON library in order to test this out - and the resulting sqlite file will break once further PRs come in that make more modifications to the schema. 

TLDR: Only library versions jumping from directly v9.4 (or earlier) -> the first v9.5 Pre-Release are supported, as we're still fixing and changing the SQL schema.

CyanVoxel commented:
> > Since this PR drastically changes elements of the schema (and is continuing to tweak things during review) you'll likely have to re-migrate your JSON library in order to test this out
> 
> That is what I did though, I am aware that the sqlite schema changed, but since this is merging to main, I imagine it still wants to support migrating from 9.4 -> this version, right?
> 
> To clarify, I am not migrating with main and then opening with this branch, both migration and testing are done with this branch, from a (JSON) library created with 9.4
> 
> But okay, what I read from this is that it's expected that this branch breaks migration and we're okay and aware of that 👍

Sorry, I assumed you were opening a library migrated from main. I'm used to an issue where I try opening a library that I migrated with main in this version but get an error in the console due to the differing columns and just assumed that's what you were getting as well.

This PR 100% should *not* break the migration code, just compatibility with libraries that were migrated before this PR.

> I didn't want to report it before because the PR was a draft, but with a build from this PR I can't open video files from tagstudio in a library migrated from 9.4 (migration also done with this build)

I'm somewhat of an idiot, this is the first time I noticed you saying "video files" and not just libraries migrated. 
Could you describe the problem you're encountering in further detail, maybe with some screenshots? Do thumbnails appear in the thumbnail grid? Does the video still play in the preview panel? Does it let you open the file itself in its default program, as well as in your file explorer? Are you trying to select more than one file at once?

CyanVoxel commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -631,22 +645,22 @@ def search_tags(
             return res
 
     def get_all_child_tag_ids(self, tag_id: int) -> list[int]:
```
Okay, so while renaming this I thought about calling it `get_ancestor_tag_ids()` if it was actually getting parents recursively given a tag, but then realized a couple things:
1. This has been something I **needed** for getting ancestor tags (I've also called it a "tag cluster" elsewhere in the code)
2. This method isn't even getting called by anything besides itself??

Here was my suggested refactor, but I honestly don't know if it's accurate to what this method is during, or what this method is even doing if anything, and whether or not this should be used elsewhere in this PR:
```py
def get_ancestor_tag_ids(self, tag_id: int) -> list[int]:
        """Recursively traverse a Tag's parent tags and return a list of all ancestor tag IDs."""
        ancestor_ids: set[int] = {tag_id}

        with Session(self.engine) as session:
            tag: Tag | None = session.scalar(select(Tag).where(Tag.id == tag_id))
            if tag is None:
                raise ValueError(f"No tag found with id {tag_id}.")

            parent_ids = tag.parent_ids

        ancestor_ids.update(parent_ids)

        for parent_id in parent_ids:
            ancestor_ids.update(self.get_ancestor_tag_ids(parent_id))

        return list(ancestor_ids)
 ```

Computerdores replied:
> 2. This method isn't even getting called by anything besides itself??

Yeah you're right. I would suggest removing it for now and should this be needed again we can add the refactored version back

CyanVoxel commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -934,59 +921,43 @@ def add_tag(
+        tag_ids_ = [tag_ids] if isinstance(tag_ids, int) else tag_ids
+        with Session(self.engine, expire_on_commit=False) as session:
+            try:
+                for tag_id in tag_ids_:
+                    tag_entry = session.scalars(
+                        select(TagEntry).where(
+                            and_(
+                                TagEntry.tag_id == tag_id,
+                                TagEntry.entry_id == entry_id,
+                            )
+                        )
+                    ).first()
+                    if tag_entry:
+                        session.delete(tag_entry)
+                        session.commit()
```
Sounds good to me, yeah I believe this is used in the bulk editing process so a more efficient query would definitely be beneficial. A bit more than I can manage right now though, so I'll add a TODO here ;)

CyanVoxel commented:
@Computerdores I've reworked the migration of edited built-in JSON tags in a way that addresses all of the comments regarding the `pass`es on the built-in range as well as the default behavior of trying to add all built-in tags every time a library is opened. The only thing I haven't done and am struggling with is trying to get JSON tag color overrides for built-in tags applying to the SQL tags. The `update_tags()` method doesn't do anything specific for this, and in fact just calls `add_tags()` in the background which is the same thing that the `buld_tag` panel uses to update colors, so I'm not sure where I'm going wrong here. This is also responsible for the failing test.

For the rest of the comments on `library.py` I believe those all hinge on the verdict of this: https://github.com/TagStudioDev/TagStudio/pull/655#discussion_r1911984781.

> completely commenting out some of the tests is a bit iffy, but I think it'll be fine

And as far as this goes, I agree that it's not ideal but I certainly wasn't expecting to have to refactor the item grid selection system to behave more like how it did in v9.4 as a part of this feature *(whose purpose was to add a future v9.6 feature in order to avoid bugs stemming from flaws in the how fields are connected to entries in the DB).* So I guess I'd consider rewriting all of those tests to be out of the scope of this PR, especially since they'll likely all need to be rewritten again at some point. In my opinion it was just too early for those tests to exist.

Tishj commented:
> Do thumbnails appear in the thumbnail grid? Does the video still play in the preview panel? Does it let you open the file itself in its default program, as well as in your file explorer? Are you trying to select more than one file at once?

I've just made a reproduction with some stock videos.
With 9.4.2 I make a library, the folder contains 4 videos (.mp4 files)

I've added some tags to them to help simulate my real library
Environment
- Forest
Season
- Winter

I close the old version, reopen with a build from this PR (haven't pulled new changes since our last conversation though)
migrate successfully, the library works correctly, thumbnails and previews are working.
What's even stranger is that I just discovered that when I right click on the preview and "Open file" from there it *does* work.
But when I right click on an entry in the main panel (search results I guess you'd call that?) and select "Open file" it doesn't work.

I just pulled `main` and build that, there both work (I went into `.TagStudio` of the library and deleted the `.sqlite` file to let it remigrate from json)

CyanVoxel commented:
> > Do thumbnails appear in the thumbnail grid? Does the video still play in the preview panel? Does it let you open the file itself in its default program, as well as in your file explorer? Are you trying to select more than one file at once?
> 
> I've just made a reproduction with some stock videos. With 9.4.2 I make a library, the folder contains 4 videos (.mp4 files)
> 
> I've added some tags to them to help simulate my real library, but I doubt that matters?
> 
> ```
> Environment
> - Forest (Environment is it's parent tag)
> Season
> - Winter (Season is it's parent tag)
> ```
> 
> I close the old version, reopen with a build from this PR (haven't pulled new changes since our last conversation though) migrate successfully, the library works correctly, thumbnails and previews are working. What's even stranger is that I just discovered that when I right click on the preview and "Open file" from there it _does_ work. But when I right click on an entry in the main panel (search results I guess you'd call that?) and select "Open file" it doesn't work.
> 
> I just pulled `main` and build that, there both methods of "Open file" work (I went into `.TagStudio` of the library and deleted the `.sqlite` file to let it remigrate from json)
> 
> In this test I am selecting a single file. The file opens in the default application when I double click it in the file explorer. The file opens in the default application when I "Open file" from the preview panel that opens to the right after selecting the entry. The file does not open in the default application (it does nothing) when I select "Open file" in the menu that opens when you rightclick on an entry in the search result panel. I also believe "show in file explorer" doesn't work from there, but that's going off of memory

I've just replicated this, however it's not limited to just videos - the "open file" right-click options just no longer work in the thumbnail grid for me. That's definitely an area I touched in this PR, so I'll give it a look. Thank you for your thorough response!
OH actually in the logs I can see what's happening, it's not getting passed the full filepath. Should be quick fix!
```
2025-01-11 08:26:41 [info     ] Opening file                   path=PosixPath('dice.jxl')
2025-01-11 08:26:41 [error    ] File not found                 path=PosixPath('dice.jxl')
```

Computerdores reviewed (approved):
Has my approval as all comments have been addressed!

Computerdores commented:
I did just notice that the tests are failing, do you know what is up with that?

Computerdores commented:
> > I did just notice that the tests are failing, do you know what is up with that?
> 
> Mentioned here: [#655 (comment)](https://github.com/TagStudioDev/TagStudio/pull/655#issuecomment-2585261121)
> 
> > The only thing I haven't done and am struggling with is trying to get JSON tag color overrides for built-in tags applying to the SQL tags. The update_tags() method doesn't do anything specific for this, and in fact just calls add_tags() in the background which is the same thing that the buld_tag panel uses to update colors, so I'm not sure where I'm going wrong here. This is also responsible for the failing test.

I have narrowed down the problem to this line: https://github.com/TagStudioDev/TagStudio/blob/tag-categories-feat/tagstudio/src/core/library/alchemy/library.py#L897
In the test that fails because of the unique constraint on `tags.id` because it is trying to insert the modified tag. What I don't understand is why that works when the UI does the same thing

CyanVoxel commented:
> > > I did just notice that the tests are failing, do you know what is up with that?
> > 
> > 
> > Mentioned here: [#655 (comment)](https://github.com/TagStudioDev/TagStudio/pull/655#issuecomment-2585261121)
> > > The only thing I haven't done and am struggling with is trying to get JSON tag color overrides for built-in tags applying to the SQL tags. The update_tags() method doesn't do anything specific for this, and in fact just calls add_tags() in the background which is the same thing that the buld_tag panel uses to update colors, so I'm not sure where I'm going wrong here. This is also responsible for the failing test.
> 
> I have narrowed down the problem to this line: https://github.com/TagStudioDev/TagStudio/blob/tag-categories-feat/tagstudio/src/core/library/alchemy/library.py#L897 In the test that fails because of the unique constraint on `tags.id` because it is trying to insert the modified tag. What I don't understand is why that works when the UI does the same thing

I managed to "fix" it by grabbing the built-in tags from the DB, editing the color, and then re-adding/"updating" the modified tags to the DB. Why this works, and why I couldn't do it in one step, I don't know...

```py
# Tags
for tag in json_lib.tags:
    self.add_tag(
        Tag(
            id=tag.id,
            name=tag.name,
            shorthand=tag.shorthand,
            color=TagColor.get_color_from_str(tag.color),
        )
    )
    # Apply user edits to built-in JSON tags.
    if tag.id in range(RESERVED_TAG_START, RESERVED_TAG_END + 1):
        updated_tag = self.get_tag(tag.id)
        updated_tag.color = TagColor.get_color_from_str(tag.color)
        self.update_tag(updated_tag)  # NOTE: This just calls add_tag?
 ```

Computerdores commented:
> > > > I did just notice that the tests are failing, do you know what is up with that?
> > > 
> > > 
> > > Mentioned here: [#655 (comment)](https://github.com/TagStudioDev/TagStudio/pull/655#issuecomment-2585261121)
> > > > The only thing I haven't done and am struggling with is trying to get JSON tag color overrides for built-in tags applying to the SQL tags. The update_tags() method doesn't do anything specific for this, and in fact just calls add_tags() in the background which is the same thing that the buld_tag panel uses to update colors, so I'm not sure where I'm going wrong here. This is also responsible for the failing test.
> > 
> > 
> > I have narrowed down the problem to this line: https://github.com/TagStudioDev/TagStudio/blob/tag-categories-feat/tagstudio/src/core/library/alchemy/library.py#L897 In the test that fails because of the unique constraint on `tags.id` because it is trying to insert the modified tag. What I don't understand is why that works when the UI does the same thing
> 
> I managed to "fix" it by grabbing the built-in tags from the DB, editing the color, and then re-adding/"updating" the modified tags to the DB. Why this works, and why I couldn't do it in one step, I don't know...
> 
> ```python
> # Tags
> for tag in json_lib.tags:
>     self.add_tag(
>         Tag(
>             id=tag.id,
>             name=tag.name,
>             shorthand=tag.shorthand,
>             color=TagColor.get_color_from_str(tag.color),
>         )
>     )
>     # Apply user edits to built-in JSON tags.
>     if tag.id in range(RESERVED_TAG_START, RESERVED_TAG_END + 1):
>         updated_tag = self.get_tag(tag.id)
>         updated_tag.color = TagColor.get_color_from_str(tag.color)
>         self.update_tag(updated_tag)  # NOTE: This just calls add_tag?
> ```

I have the strong suspicion that it is ORM shenanigans. Probably something like: The ORM notices that the Tag object that is being added is already being tracked by it and so instead of being INSERTed it gets UPDATEd instead

CyanVoxel commented:
Thank you SO much for the review, @Computerdores and @Tishj!! I know this was a massive one to look over so I *really* appreciate it!

Tishj commented:
Thanks for your continued work on the library, it's great to see the strides that the project is making! 💪
